lib/Monads/No_Fail.thy

```diff
@@ -225,4 +225,12 @@ lemma no_fail_condition:
   unfolding condition_def no_fail_def
   by clarsimp
 
+lemma no_fail_ex_pre:
+  "(\<And>x. no_fail (P x) f) \<Longrightarrow> no_fail (\<lambda>s. \<exists>x. P x s) f"
```

Comment:
Interesting. This is entirely analogous to `hoare_vcg_ex_lift`, and as such probably fine (I'd rename to `no_fail_ex_list`, though). But now, looking at it I'd have expected something like

```
"no_fail (P ?x) f ⟹ no_fail (λs. ∃x. P x s) f"
```
and the form we have here for `∀` instead. Have you tried that? Is it true?

If yes, we should potentially change `hoare_vcg_ex_lift` if that is also true in the schematic form. The `?x` might become inconvenient, not clear. We can leave that rabbit hole for a different PR/investigation, but I'd be curious if it's true at least.

Reply:
Ignore me, the scope of the `⋀` ends before the implication, so this is entirely equivalent with `P ?x` and much nicer to work with. And that's why `hoare_vcg_ex_lift` is not too strong either. All good.

I'd still rename this one to `no_fail_ex_lift`, though.